### ISSUE TITLE
Reduce missing in DDP's training/validation_step_end

### ISSUE THREAD

**Issue by apacha** [PyTorchLightning/pytorch-lightning#4853]
## 🐛 Bug

TLDR: `validation_step_end` does not aggregate the results when using DDP backend for evaluating on the entire dataset, instead of the individual splits.

The full description can be found in this SO question: https://stackoverflow.com/q/64499294/448357

It seems that the respective code might have been forgotten, as it can be found both in [DP accelerator](https://github.com/PyTorchLightning/pytorch-lightning/blob/404af43cde6696d04bb1899da2bb7e334e49716d/pytorch_lightning/accelerators/dp_accelerator.py#L140) and [DDP2 accelerator](https://github.com/PyTorchLightning/pytorch-lightning/blob/404af43cde6696d04bb1899da2bb7e334e49716d/pytorch_lightning/accelerators/ddp2_accelerator.py#L90), but not in the [DDP accelerator](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/accelerators/ddp_accelerator.py#L47).

### To Reproduce
Run training with DDP backend

### Expected behavior

One call after each epoch with the aggregated results in `validation_step_end`.

### Environment

 - PyTorch Version (e.g., 1.0): 1.0.8
 - OS (e.g., Linux): Ubuntu
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Python version: 3.8
 - GPU models and configuration: 2xNvidia 1080Ti

**Comment by tchaton** [comment]
Dear @apacha,

Thanks for using Lightning. Reduction is supported only on training_step, validation_step, test_step.
It is by design :)

Best regards,
Thomas Chaton.

**Comment by apacha** [comment]
Am I missing something? I've stated that I am indeed using `training_step_end` and `validation_step_end` as you indicated, but I am NOT getting `all_something_across_ddp` but instead just the share that was assigned to a single GPU. This happens only when I'm using DDP accelerator, but works as expected with DDP2 and DP backends. Is this really how it was intended to work? And if so, why?

**Comment by tchaton** [comment]
Hey @apacha,

Let me try it out on my side as it should work.

Best regards,
Thomas Chaton.

**Comment by williamFalcon** [comment]
validation_step_end is ONLY for DP and DDP2...

It's meant to aggregate parts of the batch. In DDP there are no parts of the batch because each process runs on its own GPU.

```python
# dp
num_gpus = 4
batch_parts = split_batch(batch, num_gpus)
model_copies = copy(model, num_gpus)

all_batch_outs = []
for batch in data:
    # run 1 batch
    batch_outs = []
    for model, batch_part in zip(model_copies, batch_parts):
        subbatch_out = model.validation_step(batch_part)
        batch_outs.append(subbatch_out)

    reduced = validation_step_end(batch_outs)
    all_batch_outs.append(reduced)

validation_epoch_end(all_batch_outs)
```

The documentation has videos that explain all this

**Comment by apacha** [comment]
May I ask what you would recommend me to do? I have to use DDP because the other backends are creating issues in my (hardware) environment. My problem is that validation only ever sees 1/4th of my validation set when training on four GPUs. But I would like to evaluate on the complete validation set. At least I thought that `validation_step_end`'s intention was exactly that, but maybe I'm wrong.

I'll summarize: I want to run validation on all results from all GPUs to obtain aggregated statistics. How would I do that with the DDP backend?

I read through the [documentation](https://pytorch-lightning.readthedocs.io/en/stable/multi_gpu.html) and watched your video again, but I could not find any answer to that question. That's why I also posted it on Stackoverflow, where I also got the feedback, that I should used `validation_step_end`.

**Comment by tchaton** [comment]
Hey @apacha,

There is 2 options. 
1. You could create a metric using our Metric API which will take care of the reduction across DDP (recommended).
2. In 1.1.0, we introduced a self.gather function in the LightningModule to enable this use case.
Here is the function: https://github.com/PyTorchLightning/pytorch-lightning/blob/4ebce38478f28c70edc2c7236c514665df105217/pytorch_lightning/core/lightning.py#L368

Best regards,
T.C

**Comment by liutianlin0121** [comment]
>     2\. torch

Hi! May I ask how to use the second option with self.all_gather? I wish to gather all validation results after a epoch and average across them, and to do this I have naively tried
 
```
    def validation_epoch_end(self, validation_step_outputs):
        self.all_gather(validation_step_outputs)
        self.log('val_metric', sum(validation_step_outputs)/ len(validation_step_outputs),  prog_bar = True)
```

But this does not work. I wonder how to properly use self.all_gather in this case?

**Comment by liutianlin0121** [comment]
@apacha May I ask if you have solved this problem in the end? I met the exactly same problem for ddp.

**Comment by apacha** [comment]
No, I haven't solved it yet. That's why I created this ticket. Once I have time, I will try the Metric API approach

**Comment by MarsSu0618** [comment]
Execute me. 
May I ask if it have solved this problem in the end? I met the exactly same problem for ddp, too.
Have other solution?

**Comment by Vedaad-Shakib** [comment]
See this example for usage, which applies to `training_step_end` and `validation_step_end` as well: https://github.com/Lightning-AI/lightning/discussions/9259#discussioncomment-1858041
